### PR TITLE
[Keymap][user] user dvorak_42_key ergodox_ez and crkbd keymaps

### PR DESCRIPTION
**Pull request by luc-languagetools**
## Description
Add user dvorak_42_key in userspace. add crkbd keymap.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/crkbd/keymaps/dvorak_42_key/keymap.c [hunk 1]
```diff
@@ -0,0 +1,287 @@
+
```
Comment (waffle87) [inline keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:1]: All source files need a GPL2+ license header:
```
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/keymaps/dvorak_42_key/keymap.c (hunk 1 above)
Comment (luc-languagetools) [reply keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:4]: done, thank you.

File: keyboards/crkbd/keymaps/dvorak_42_key/config.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2022 LucW (@luc-languagetools)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+
+#pragma once
+
+//#define USE_MATRIX_I2C
```
Comment (drashna) [inline keyboards/crkbd/keymaps/dvorak_42_key/config.h:7]:
```suggestion
```

File: keyboards/crkbd/keymaps/dvorak_42_key/keymap.c [hunk 3]
```diff
@@ -0,0 +1,290 @@
+// Copyright 2022 LucW (@luc-languagetools)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include "dvorak_42_key.h"
+#include "rgblight.h"
```
Comment (drashna) [inline keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:8]: Shouldn't be needed:
```suggestion
```

File: keyboards/crkbd/keymaps/dvorak_42_key/keymap.c [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  if (!is_keyboard_master()) {
+    return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
+  }
+  // return rotation;
+  return OLED_ROTATION_180;
+}
+  
+
+#define DISPLAY_LAYER_NAME(LAYER_NAME, LAYER_STRING) \
+  if(biton32(layer_state) == LAYER_NAME) { \
```
Comment (drashna) [inline keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:156]:
```suggestion
  if(get_highest_layer(layer_state) == LAYER_NAME) { \
```

File: users/dvorak_42_key/config.h [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2022 LucW (@luc-languagetools)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#define DOUBLE_TAP_SHIFT_TURNS_ON_CAPS_WORD
```
Comment (drashna) [inline users/dvorak_42_key/config.h:4]:
```suggestion
#pragma once

#define DOUBLE_TAP_SHIFT_TURNS_ON_CAPS_WORD
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by luc-languagetools** [comment]
note: the latest CI build error mentions the firmware is too big:
```
Checking file size of crkbd_r2g_dvorak_42_key.hex
 * The firmware is too large! 28918/28672 (246 bytes over)
 [ERRORS]
```
however when I build locally, I get the following result:
```
The firmware size is fine - 25776/28672 (89%, 2896 bytes free)
```

**Comment by waffle87** [comment]
> note: the latest CI build error mentions the firmware is too big:

I assume the CI uses avr-gcc 5.x, which results in larger binaries, while you probably have 8.x. You can see if there's anything that can be trimmed — [squeezing AVR](https://docs.qmk.fm/#/squeezing_avr), but I'd say it isn't too big of a deal.

File: users/dvorak_42_key/dvorak_42_key.h [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                              T(A),\
+                              T(S),\
+                              T(T),\
+                              T(E),\
+                              T(SPC),\
+							  T(NUM),\
+							  T(ENTER),\
+                             END);\
+            }\
+        break;\
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.h:183]:
```suggestion
```
These are for the old macro system which has been removed and will not work.

File: users/dvorak_42_key/dvorak_42_key.c [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            }
+    }
+
+  // shell macros
+  if(record->event.pressed) {
+    switch (keycode) {
+        case SHELL_LS:
+            SEND_STRING("ls\n");
+            return true;
+            break;
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:24]: `break` after `return` is dead code.

File: users/dvorak_42_key/dvorak_42_key.h [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+#define COMBINED         6 // combined numbers and symbols layer
+#define ANDROID_STUDIO   7 // android studio specific layer
+#define VSCODE           8 // visual studio code specific layer
+#define SHORTCUTS        9 // shortcuts to be intercepted by autohotkey
+
+enum custom_keycodes {
+  PLACEHOLDER = SAFE_RANGE, // can always be here
+  EPRM,
+  VRSN,
+  RGB_SLD,
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.h:26]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Nope, CI uses AVR-GCC 8. The problem is likely due to using a lot of rather heavy features (tap dance, caps word, RGB, OLED, etc), and a fair amount of layers. So you may have to remove some stuff.

**Comment by luc-languagetools** [comment]
I notice that the "rev1" build succeeds but not the r2g. Is there a way to declare that my keymap only builds on rev1?

**Comment by fauxpark** [comment]
You can move your keymap folder under the rev1 keymaps folder (create it if it doesn't exist).

**Comment by luc-languagetools** [comment]
@drashna thank you for your time. i've implemented your changes and i've moved the keymap to `keyboards/crkbd/rev1/keymaps/dvorak_42_key`. Does this change now meet the pre-requisite to be merged ?

File: users/dvorak_42_key/dvorak_42_key.c [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        case SHELL_GIT_DIFF:
+            SEND_STRING("git diff\n");
+            break;
+
+        case SHELL_GIT_STATUS:
+            SEND_STRING("git status\n");
+            break;            
+
+        case WINDOWS10_WORKSPACE_LEFT:
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_LEFT))));
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:65]:
```suggestion
            tap_code16(C(G(KC_LEFT)));
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case WINDOWS10_WORKSPACE_RIGHT:
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_RIGHT))));
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:68]:
```suggestion
            tap_code16(C(G(KC_RIGHT)));
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case WINDOWS10_TASK_VIEW:
+            SEND_STRING(SS_LGUI(SS_TAP(X_TAB)));
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:71]:
```suggestion
            tap_code16(G(KC_TAB));
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            SEND_STRING(SS_LCTL("a") "n");
+            break;
+        case SCREEN_NEW_TAB:
+            SEND_STRING(SS_LCTL("a") "c");
+            break;
+        case SCREEN_DETACH:
+            SEND_STRING(SS_LCTL("a") "d");
+            break;
+        case SCREEN_RENAME:
+            SEND_STRING(SS_LCTL("a") SS_LSFT("a"));
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:87]:
```suggestion
            SEND_STRING(SS_LCTL("a") "A");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+            SEND_STRING(SS_LCTL("a") "9");
+            break;
+        case SCREEN_COPY_MODE:
+            SEND_STRING(SS_LCTL("a") "[");
+            break;
+        case SCREEN_PASTE:
+            SEND_STRING(SS_LCTL("a") "]");
+            break;
+        case SCREEN_KILL:
+            SEND_STRING(SS_LCTL("a") "k" "y");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:129]:
```suggestion
            SEND_STRING(SS_LCTL("a") "ky");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+            break;            
+        case SCREEN_READREG_1:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 1\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:134]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 1\n");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case SCREEN_READREG_2:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 2\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:139]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 2\n");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case SCREEN_READREG_3:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 3\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:144]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 3\n");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+
+        case SCREEN_PASTEREG_1:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 1\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:150]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 1\n");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case SCREEN_PASTEREG_2:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 2\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:155]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 2\n");
```

File: users/dvorak_42_key/dvorak_42_key.c [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+            break;
+        case SCREEN_PASTEREG_3:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 3\n");
```
Comment (fauxpark) [inline users/dvorak_42_key/dvorak_42_key.c:160]:
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 3\n");
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by luc-languagetools** [comment]
@drashna can this change be merged ? I believe i've implemented all of the requested changes.

**Comment by luc-languagetools** [comment]
Hi, can this be merged ? Please let me know. Thank you.

**Review by drashna — APPROVED** [review]

**Comment by luc-languagetools** [comment]
Thank you !

**Pull request merged**